tls/s2n_record_read.c

```diff
@@ -174,8 +175,23 @@ int s2n_record_parse(struct s2n_connection *conn)
         iv.size = sizeof(aad_iv);
 
         GUARD(s2n_stuffer_init(&iv_stuffer, &iv));
-        GUARD(s2n_stuffer_write_bytes(&iv_stuffer, implicit_iv, cipher_suite->record_alg->cipher->io.aead.fixed_iv_size));
-        GUARD(s2n_stuffer_write_bytes(&iv_stuffer, en.data, cipher_suite->record_alg->cipher->io.aead.record_iv_size));
+
+        if (cipher_suite->record_alg->flags & S2N_TLS12_AEAD_NONCE) {
+            GUARD(s2n_stuffer_write_bytes(&iv_stuffer, implicit_iv, cipher_suite->record_alg->cipher->io.aead.fixed_iv_size));
+            GUARD(s2n_stuffer_write_bytes(&iv_stuffer, en.data, cipher_suite->record_alg->cipher->io.aead.record_iv_size));
+        } else if (cipher_suite->record_alg->flags & S2N_TLS13_AEAD_NONCE) {
```

Comment:
The naming is probably too optimistic. This is the nonce construction specified in https://tools.ietf.org/html/rfc7905 for the TLS1.2 ChaCha suites. A re-read of https://tlswg.github.io/tls13-spec/#rfc.section.5.3 , shows this code will only work for a 96 bit IV. Shall we remove any TLS1.3 references and just call it `S2N_TLS12_CHACHA_NONCE` ?


For validation, we have the integration tests w/ GnuTLS and s_client.